src/util.cc

```diff
@@ -423,21 +427,79 @@ const char* SpellcheckString(const char* text, ...) {
 }
 
 #ifdef _WIN32
+wstring Utf8ToWide(const string& u8_string) {
+  if (u8_string.empty())
+    return wstring();
+
+  size_t size_needed = MultiByteToWideChar(CP_UTF8, 0, u8_string.data(),
+                                           (int)u8_string.size(), NULL, 0);
+  if (size_needed == 0)
+    Fatal("Failed conversion to wide string");
+
+  vector<wchar_t> w_buffer(size_needed);
+  int w_size_needed =
+      MultiByteToWideChar(CP_UTF8, 0, u8_string.data(), (int)u8_string.size(),
+                          &w_buffer[0], w_buffer.size());
+  if (w_size_needed == 0)
+    Fatal("Failed conversion to wide string");
+
+  return std::wstring(&w_buffer[0], w_size_needed);
+}
+
+string WideToUtf8(const wstring& w_string) {
+  if (w_string.empty())
+    return std::string();
+
+  int size_needed = WideCharToMultiByte(
+      CP_UTF8, 0, &w_string[0], (int)w_string.size(), NULL, 0, NULL, NULL);
+  if (size_needed == 0)
+    Fatal("Failed conversion to UTF-8 string");
+
+  std::string u8_string(size_needed, 0);
+  WideCharToMultiByte(CP_UTF8, 0, &w_string[0], (int)w_string.size(),
+                      &u8_string[0], size_needed, NULL, NULL);
+  if (size_needed == 0)
+    Fatal("Failed conversion to UTF-8 string");
+
+  return u8_string;
+}
+
+char** convertCommandLine(int argc, wchar_t** wargv) {
```

Comment:
I'm really wary about this function. It seems like then wrong direction.

I'd rather see the parts of Ninja that actually care about commandline arguments coming in as argc / wargv be modified to accept a template type of CHAR_T, so that the commandline parameters can be passed in directly, and then after that parsing is completed, we never access argc / argv again.

That way you avoid the memory leak problem entirely.

Reply:
But where would the conversion to UTF-8 happen?

Reply:
Any of the commandline arguments that need to be stored internally as strings would be converted during the cli parsing.

Anything that is stored as something other than a string (e.g. -j number) would be stored the way it's currently stored.

This also avoids the need to convert string data that doesn't get kept after cli-parsing, such as long-form option names.

Reply:
I'm not sure I understand, but @jlonnberg hopefully does.

Reply:
I'll try to elaborate more.

Imagine this is how Ninja currently does things (At this moment, I have not checked the code to confirm this. I'm just giving this as an example)

```cpp
struct CommandlineOptions
{
    // Various options, such as
    std::size_t numJobs;
    std::string pathToNinjaFile;
    // and so on.
}

std::string convertToUTF8String(char* str)
{
    return std::string(str);
}

std::string convertToUTF8String(wchar_t* str)
{
    return WideStrToUTF8(str);
}

template<typename CHAR_T>
CommandlineOptions parse_cli_options(std::size_t argc, CHAR_T ** argv)
{
    CommandlineOptions args;
    for(/* each option */)
    {
        if(option == "-j")
        {
            args.numJobs = parse(data);
        }
        else if(option == "-- some string option")
        {
                args.someStringOption = convertToUTF8String(data);
        }
    }  
}

int main(argc, argv)
{
    CommandlineOptions args = parse_cli_options(argc, argv);
    the_rest_of_ninja(args);
}
```

With a convention like the above, whether or not the arguments passed to the program by the operating system are char or wchar_t is only relevant when parsing the CLI arguments.

Once the CLI arguments are parsed, the rest of the codebase does not care.

This way, we avoid, entirely, the need to allocate a new char[][]. and a new char[] for each argument, which means less work for Ninja to do.

Reply:
I see. I would prefer not to touch the other platform's code so that the situation on Windows remains what it is: shit.

Reply:
You prefer not to touch other platforms code for this PR? or at all?

Reply:
I prefer not to touch other platforms code to work-around Windows problems.

Reply:
I don't get it.

This pattern of parsing commandline arguments is the correct way to do it, regardless of whether this PR is the reason why it's becoming obvious.

Reply:
Note that this also becomes relevant in the "ninja as a library" concept.

Passing raw '\0' terminated char**'s into a library function is disgusting.

So in order to inform Ninja of what the library consumer wants it to do, the Ninja library will need some entry point that accepts a list of arguments in some form.

That list will either be a struct, a map, or a list of function call parameters.

I would prefer a struct, since it makes it easier to keep type safety.

Reply:
Once the function for creating the struct / map of parameters to Ninja exists, making it generic to `char` vs `wchar_t` is trivial.

Reply:
Out-of-scope for this PR though.

Reply:
Is it really?

You'd rather have the Windows build of Ninja adopt such a crappy way of doing things as allocating a char**, and copying the wchar_t** over into it, instead of just fixing the problem for everyone?

If you *really* want to keep the two concepts separate, then I recommend considering a PR that implement appropriate commandline parsing as a prerequisite of merging this PR.

Reply:
If it's so easy to "just" fix the problem for everyone, I wonder why it hasn't been done yet.

Reply:
Because of these reasons:

* there are over 60 pull requests open (most of them several years old), and before you started chugging on the Ninja project it was over 100 of them (good work, of course, but lets not let a recent improvement detract from historical context).
* The Ninja project has historically refused to allow people to use C++11, which drastically simplifies so many tedious tasks, though now that restriction has been lifted for Windows builds.
* You're arguing that it's not a better way to do it, therefore I'm not going to bother writing the code for it just to see it languish in perpetuity.
* Ninja uses the "getopt" library (embedded into the ninja source code), which has been patched to support esoteric platforms like AIX, and I feel that that you and/or Nico will expect absolute bug-for-bug compatibility (if any bugs exist, i suppose) with the existing getopt, in any rewritten cli parser. But please correct me if I'm mistaken.
* This PR is the first PR that I'm aware of that attempts to add support for Windows's broken API decision of natively using wchar_t (and not even the same wchar_t as other platforms, of course).
* Previous discussions with Nico about having better support for multiple methods of giving Ninja configuration data (environment variables, commandline, configuration files) have resulted in outright rejection, leaving no user-facing feature improvements to provide motivation to work on something like this.

Just like you, I'm a volunteer, and so are many (not all, but many) of the people who submit PRs.

It *used* to be my job to work on Ninja related things, but not any more. So given all of the ways I can predict a PR to implement this would be rejected, I'm not going to spend my free time making one.

But if you can give me some assurances that it wouldn't be a waste of my time, then I can allocate some time to work on it.

Reply:
> [...] though now that restriction has been lifted for Windows builds.

There's no "restriction" that's been lifted. Especially not by "the Ninja project".

> You're arguing that it's not a better way to do it

Not true.

> But if you can give me some assurances that it wouldn't be a waste of my time, then I can allocate some time to work on it.

Can't give you that, sorry.

Reply:
> There's no "restriction" that's been lifted.

Seriously dude? Don't lie.

Quote Nico: There are no plans on adopting C++11.
https://github.com/ninja-build/ninja/issues/1403
Whereupon followed an extremely lengthy discussion about how Ninja wouldn't accept PRs that used C++11 code.

Quote you: set -std=c++98 in cflags
https://github.com/ninja-build/ninja/pull/1566

Reference: Something on the order of 25%+ of the discussions that you and I have had in various issues and PR's here on github since january 2019.

And then quote you here, in this PR, that Windows may use C++11.

>  Especially not by "the Ninja project".

You *ARE* the Ninja project. You are the gatekeeper for every improvement that is made to this codebase for the last year or more. 

Nico has had *nearly* no activity, either with code, or PR comments, or really anything for well over a year. Yes, there's been the rare action from them, but nothing substantive.

I'm unable to find any evidence of Evmar doing anything for the last year. I would be happy to be shown where my search skills are lacking.

So unless you're communicating with the other listed maintainers outside of the mailing list, or github, it's disingenuous to claim that there's anyone making decisions about these things other than you.

> > You're arguing that it's not a better way to do it

> Not true.

Honestly, it really seems like you are.

So if I'm reading your comments in a different way than you intended me to, I apologize.

> Can't give you that, sorry.

This, right here, is why people haven't already made a PR to do something that's "so easy" 

You want your cake, and you want to eat it to.

It's incredibly frustrating to work with this project for exactly this reason. Attempts at improvements to the codebase sometimes sit for *years* with no comments, while other PRs are merged within days.

We can't officially move to anything resembling modern C++ because we have some *unspecified* list of reasons that no one can provide, "something something, old platforms".

And the one person with "Collaborator" or "Maintainer" or "Owner" next to their name, who actually does any work, can't provide any guarantees about a chunk of work being discussed, so that the volunteer who's saying they can work on it doesn't feel like they're about to waste a bunch of their time.

You could have said:

I promise that if you write a PR for this, I will discuss it with you promptly, and we will try to make some kind of improvement to the code, even if the initial draft isn't acceptable.

Or you could have said:

That *does* sound like a good way to improve the code, but until I see it I cant make any promises about whether its something that I will merge, because we might find other problems that make it a bad idea.

Or at the absolute minimum you could have said:

All I can promise is that I won't outright reject it without a thorough review and proper discussion with you, and at a very minimum it'll be beneficial to the project because of all of the analysis we do but I'm not sure we'll merge it one way or another because it's not clear it's the right design.

But no, you just said: 

> Can't give you that, sorry.

So yea, my time's not valuable. Thanks for letting me know.

Reply:
> And then quote you here, in this PR, that Windows may use C++11.

I didn't say that. You're paraphrasing a lot, that's why I need to disagree, so that you don't quote me next time with "here you stated officially that the Ninja project wants to port the code to C++11 on Windows" or something like that.

> This, right here, is why people haven't already made a PR to do something that's "so easy"

Why are there *any* PRs there then? And some are definitely for stuff that's easier.

Reply:
> I didn't say that. You're paraphrasing a lot, that's why I need to disagree, so that you don't quote me next time with "here you stated officially that the Ninja project wants to port the code to C++11 on Windows" or something like that.

It's a binary choice. Either C++11 may not be used on windows, or it may. Either position is the "official" position of the project.

Why can't you just pick one?

Don't make me have to interpret what you write. Don't use ambiguous wording. Just say things plainly.

So which is it? What's the policy?

> Why are there any PRs there then?

Sometimes the pain of doing something is worth doing, if the outcome is sufficiently important.

Since the change I described here in this comment chain wouldn't result in an outcome that's end-user facing, and I can't have any kind of soft promise that working on it won't be a waste of my time, the pain of doing this particular task is not worth doing.

>  And some are definitely for stuff that's easier.

I don't understand why you say this. It weakens your point and strengthens my own.

Reply:
> > And then quote you here, in this PR, that Windows may use C++11.

> I didn't say that. You're paraphrasing a lot, that's why I need to disagree, 

https://github.com/ninja-build/ninja/pull/1671#discussion_r345970019

> You can use C++11 on Windows, I don't think we support older compilers there or will support them by the time this gets merged.

How else am I supposed to read this?

Reply:
It work's like this:

I review this PR, check CI, test it on Windows (with Visual Studio 2019), and wait for some other people to give feedback.
Now if someone comments something like "I tried to compile it with Visual Studio 2010 and it didn't seem to work. Sorry, we still have some older machines here that require that" I wouldn't answer "Sorry dude, it's Ninja's official policy to use C++11, you're on your own.".
Also if any owner objects or reverts this PR, I won't argue. *I* don't make any official statements, they do. So the last official statement is still "Ninja doesn't use C++11".
It seems likely to me though that neither scenarios won't happen, that's why I said what I said.

> So unless you're communicating with the other listed maintainers outside of the mailing list, or github, [...]

We are.

Reply:
> We are.

How about you all share with the class then?

Reply:
It isn't much and anything important is public. There's also an IRC channel btw, don't know if you consider that public as there are no logs (but anyone can join).

Reply:
> Now if someone comments something like "I tried to compile it with Visual Studio 2010 and it didn't seem to work. Sorry, we still have some older machines here that require that" 

The proper response to this statement is

"We don't support ancient compiler tool chains."

> Also if any owner objects or reverts this PR, I won't argue.

How about they actually do some legwork then?

>  I don't make any official statements, they do. So the last official statement is still "Ninja doesn't use C++11".
> It seems likely to me though that neither scenarios won't happen, that's why I said what I said.

Forgive me for using a rude phrase, but I actually did some searching on Google for alternative ways to say this without using somewhat vulgar language.

Shit or get off the pot.

Reply:
Anyway, this has gone too offtopic. For this PR this is fine for now, let's see what the performance is like after the other remarks have been fixed.